### DISCUSSION TITLE
How can I  recover the nrf-bootloader?

### DISCUSSION THREAD

**Discussion by shang-huan** [bitcraze/discussions#425] [Q&A] [answered]
For some reasons, I want to recover the nrf-bootloader, but I'm not sure how to do. Is it the same as  the recovering the 
stm-bootloader operation?

**Comment by ataffanel** [comment, marked as answer]
Hi!

It might be interesting to know more about the reason why this recovery is needed: it should not be possible (ie. very hard) to brick the nrf bootloader over radio, so I am curious of how that could happen.

The way to re-flash the bootloader is to connect to the SWD debug port on the right side of the Crazyflie, the debug-adapter-kit contains a small PCB that is designed to be soldered on the Crazyflie PCB, otherwise you could use pogo-pins or solder your own wires. Then you can run ```make factory_reset``` in the [crazyflie2-nrf-firmware repos](https://github.com/bitcraze/crazyflie2-nrf-firmware/blob/c3fc80fe1f2a94e65bed0ca4836d0b8270ae0fe1/Makefile#LL207C9-L207C9). This will re-flash all the required firmware including the Nordic bluetooth softdevice and the bootloader. By default the makefiles expects a stlink-v2 debug probe but that can be changed at the beginning of it, any probe supported by openocd should work.

Ps. I moved your question as a discussion as it is a more suited place for it. We are trying to reserve issues for actual change to the  repos code/documentation.

**Reply by shang-huan** [reply]
Thanks for your reply! When I modify the firmware code, it takes too much time to flash one by one if there are a lot of UAVs. So I want to modify the nrf-bootloader code and the stm-bootloader code to make the UAVs flash together. From your documentation, I only found out about stm-bootloader flashing content but not nrf-bootloader flashing content. I don't know how to overflash the nrf-bootloader.

**Reply by ataffanel** [reply]
This sounds very interesting. Then what you really want is to work in the crazyflie2-nrf-bootloader project and ```make flash``` to flash the bootloader. The rest of the explanation about where to connect the debug probe is the same.

Flashing a swarm all at once is something I would like to get working at some point so I would be very interested to know how it went and if you manage to get something working.

Do not hesitate to ask questions in this forum if you have any about the bootloader code or architecture.

**Reply by shang-huan** [reply]
Thanks for your answer, it helped me a lot.

**Reply by knmcguire** [reply]
@shang-huan could you mark it as an answer? Or perhaps provide a summary in 'suggest an answer' and mark that as an answer? Thanks!

**Reply by shang-huan** [reply]
Ok, I will make a summary as an answer after I try it.
